staging/src/k8s.io/apiserver/pkg/endpoints/discovery/version_hack.go

```diff
@@ -17,29 +17,75 @@ limitations under the License.
 package discovery
 
 import (
+	"net/http"
 	"regexp"
 	"sort"
 	"strings"
 
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime/schema"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
 )
 
 // HACK: support the case when we can add core or other legacy scheme resources through CRDs (KCP scenario)
-var ContributedResources map[schema.GroupVersion]APIResourceLister = map[schema.GroupVersion]APIResourceLister{}
+var ContributedResources map[ClusterGroupVersion]APIResourceLister = map[ClusterGroupVersion]APIResourceLister{}
 
-func withContributedResources(groupVersion schema.GroupVersion, apiResourceLister APIResourceLister) APIResourceLister {
-	return APIResourceListerFunc(func() []metav1.APIResource {
-		result := apiResourceLister.ListAPIResources()
-		if additionalResources := ContributedResources[groupVersion]; additionalResources != nil {
-			result = append(result, additionalResources.ListAPIResources()...)
-		}
-		sort.Slice(result, func(i, j int) bool {
-			return result[i].Name < result[j].Name
-		})
+type ClusterGroupVersion struct {
```

Comment:
Hrm, I'm just looking at this hack now (I think i glossed over it before).  Why do you need a global variable for ContributedResources that is cluster scoped?  Why not calculate this on demand from the request context?  

In the long run I could see the context carrying an injected interface that lets us perform some of these cluster scoped behaviors (where you would call methods on it instead of having lots of little caches everywhere).  You might just want to skip having the cache and do it dynamically.